### PR TITLE
Fix libz on kindle4

### PR DESCRIPTION
**Pull request by onde2rock**
I have this problem where covermenu can't find z.

This commit fix it, and work on sdl, but I have no idea about other device.

What's the best way to go about this ?

### CODE REVIEW

File: plugins/coverbrowser.koplugin/xutil.lua [hunk 1]
```diff
@@ -13,7 +13,7 @@ int compress2(uint8_t *dest, unsigned long *destLen,
 int uncompress(uint8_t *dest, unsigned long *destLen,
                const uint8_t *source, unsigned long sourceLen);
 ]]
-local zlib = ffi.load(ffi.os == "Windows" and "zlib1" or "z")
+local zlib = ffi.load(ffi.os == "Windows" and "zlib1" or "libs/libz.so.1")
```
Comment (Frenzie) [inline plugins/coverbrowser.koplugin/xutil.lua]: z is short for libz(.so/.dynlib/.dll/etc) and should be kept.

What is the precise error message? If it came from here then that's where it should be fixed:

https://github.com/koreader/koreader/blob/e21fe5504299f2c650939da22adefb73aa4f0f2b/setupkoenv.lua#L19-L31

Also pinging @poire-z who wrote the zlib stuff here.

Comment (Frenzie) [reply plugins/coverbrowser.koplugin/xutil.lua]: > z is short for libz(.so/.dynlib/.dll/etc) and should be kept.

By this I mean that at the very least what you wrote should be `or "z" *or* "new_addition"` even if it'd still be the wrong place.

**Comment by poire-z** [comment]
I have no real idea about how library loading works.
I copied and pasted that line from http://luajit.org/ext_ffi_tutorial.html, and it just worked...

If it is to be fixed here, it should also be fixed in https://github.com/koreader/koreader-base/blob/7cb11c4f52787207264d2284213240b4c065e27b/ffi/zipwriter.lua#L19-L20.
(I guess saving a wikipedia article as EPUB should then also fail on kindle4).

But I agree it looks fine with "z", and there's something wrong in ffi.load...

There is also a libs/libz.so.1 in koreader for kobo.
Looking at the luajit process, it seems it uses both the koreader one and the system one:
```
2b401000-2b402000 rw-p 00010000 b3:03 754        /mnt/onboard/.adds/koreader/libs/libz.so.1
2cd0f000-2cd21000 r-xp 00000000 b3:01 1516       /lib/libz.so.1.2.4
```

**Comment by onde2rock** [comment]
The source code for package.searchpath is straightforward, not more to expect on this side : 

```
static const char *searchpath (lua_State *L, const char *name,
                                             const char *path,
                                             const char *sep,
                                             const char *dirsep) {
  luaL_Buffer msg;  /* to build error message */
  luaL_buffinit(L, &msg);
  if (*sep != '\0')  /* non-empty separator? */
    name = luaL_gsub(L, name, sep, dirsep);  /* replace it by 'dirsep' */
  while ((path = pushnexttemplate(L, path)) != NULL) {
    const char *filename = luaL_gsub(L, lua_tostring(L, -1),
                                     LUA_PATH_MARK, name);
    lua_remove(L, -2);  /* remove path template */
    if (readable(filename))  /* does file exist and is readable? */
      return filename;  /* return that file name */
    lua_pushfstring(L, "\n\tno file '%s'", filename);
    lua_remove(L, -2);  /* remove file name */
    luaL_addvalue(&msg);  /* concatenate error msg. entry */
  }
  luaL_pushresult(&msg);  /* create error message */
  return NULL;  /* not found */
}
```

According to best practice, `libz.so` should be a symlink to `libz.so.1`.
Look like It's a problem on my device then. I can quickly fix it by do it manually. But why it's working for all other lib ?

**Comment by Frenzie** [comment]
> According to best practice, libz.so should be a symlink to libz.so.1.

Blame Microsoft (FAT). Symlinks? Hah! What a luxury. :-)


> Edit: Actually, z is the only case of using searchpath in koreader. The other ffi.load are directly calling libs/something.so.3.

Code links are always helpful. There's no need to duplicate effort. ;-)

But generally speaking it indeed looks like we're doing this over and over and over:

https://github.com/koreader/koreader-base/blob/b73bfe5f8b9e94c27d8af6272ae39a536347dc96/ffi/freetype.lua#L14-L20

https://github.com/koreader/koreader-base/blob/14fcafe8d508a8cef4fa08d5c1f4a6cc6a0a0d24/ffi/pic.lua#L8-L17

I think we should abstract that away. The pattern is always the same. Now perhaps we don't want to be too clever about it, in which case I think we should abstract it as a simple table along these lines.

```lua
local lib_names = {
  Windows = "mystery.dll",
  Linux = "mystery.so",
  OSX = "mystery.dylib",
}

local ffi = require('ffi')
ffi.load_abstraction("path?", lib_names[ffi.os])
```

**Comment by Frenzie** [comment]
> Edit: Actually, z is the only case of using searchpath in koreader. The other ffi.load are directly calling libs/something.so.3.

That doesn't mean it's a good idea to do it quite like that, imho. But it does mean it's not inside the scope of this PR unless you want it to be.

**Comment by Frenzie** [comment]
I searched around a little and over at luvit they have some cool stuff to deal with this: https://github.com/luvit/luvit/blob/9a7708469f88b1d8a6c5ee7b5910ce78e0163aa4/deps/require.lua#L245-L253

What you do there is not unlike my proposal above, but you'd pass ffi.load as the action in `Module:action()`. Anyway, that's just collecting ideas as part of the brainstorm.

The important thing about this PR as it stands is that it shouldn't break Mac OS X, which your original commit would've, so it's probably good enough for now.

**Comment by onde2rock** [comment]
Yes, I really don't want to mess with library and path since it's already working.

I also thought of doing something like this to be a little more generic in setupkoenv : 
```
-- patch ffi.load for thirdparty luajit libraries
ffi.load = function(lib)
    local loaded, re = pcall(ffi_load, lib)
    if loaded then return re end

    local lib_path = package.searchpath(lib, "./lib?.so;./libs/lib?.so")
    if lib_path then
        return ffi_load(lib_path)
    end
    for i=1,15 do
        lib_path = package.searchpath(lib, "./lib?.so."..i";./libs/lib?.so."..i)
        if lib_path then
            return ffi_load(lib_path)
        end 
    end 
    error('Not able to load dynamic library: ' .. lib)
end

```

**Comment by Frenzie** [comment]
> it's already working well enough.

Within certain contexts that we care the most about. :-)

> I also thought of doing something like this

That'd miss all of these, but yes, I think that's a reasonable enough principle.

libcrypto.so.1.0.0
libdjvulibre.so.21
libpng16.so.16
libssl.so.1.0.0

Shall I merge it as is or do you want to play around with it some more first?

**Comment by onde2rock** [comment]
> libcrypto.so.1.0.0
> libdjvulibre.so.21
> libpng16.so.16
> libssl.so.1.0.0

Ok I give up :-)

Maybe one day, someone will fix all this, but not today ...

**Comment by Frenzie** [comment]
So merge? :-)

**Comment by onde2rock** [comment]
yes please

**Pull request merged**